### ISSUE TITLE
HLSL: Is it possible to change the coordinate system?

### ISSUE THREAD

**Issue by dankbaker** [KhronosGroup/glslang#494]
D3D (e.g. HLSL) and Vulkan seem to have flipped Y cords. I was wondering what exactly ExecutionModeOriginUpperLeft and ExecutionModeOriginLowerLeft did. If I was to switch to LowerLeft would this make SPIRV produced shaders from HLSL behave like they would in D3D? If that's the case, can we change this setting for HLSL source shaders?

**Comment by johnkslang** [comment]
SPIR-V for Vulkan (but not for OpenGL) requires Execution Mode of **OriginUpperLeft**, which means both Vulkan and D3D mode. 

If we have that backwards somewhere, it should be corrected.

**Comment by dankbaker** [comment]
Y is inverted from D3D, Looks like it can be fixed by negating y in the vertex shader output but still investigating the exact difference.

My understanding was this was an understood and unfortunate difference. I was checking to see if there was a way to address it at the HLSL->Vulkan layer, but it looks like the only way to do that would be to negate the y coord via code injection then.

**Comment by johnkslang** [comment]
> Y is inverted from D3D

Inverted relative to what?

> HLSL->Vulkan layer

What's the HLSL -> Vulkan layer?

**Comment by dankbaker** [comment]
HLSL->Vulkan = glslang in HLSL mode

TLDR is that things will render upside down if you take the same shaders and run them on Vulkan compared to D3D.

**Comment by johnkslang** [comment]
Aren't there several places this could occur though, like in setting up the viewport?  I'm unclear on how we know it was the shader coordinate system.

**Comment by dankbaker** [comment]
According to baldurk it is a simple invertion on the y coordinate from D3D, and that's what it looks like to me as well. I think this has been seen by a observed by a number of people and there is a proposed extension to Vulkan.  It was my understanding that this is a well documented and known difference. (and possibly an oversight).

I don't believe there is anywhere to switch this not in the pipeline. Alot of hardware use an actual viewport matrix to do this, which means you can stuff negative values there and get whatever coordinates you want, but Vulkan doesn't take an arbitrary matrix so there is no way to change it there.

**Comment by baldurk** [comment]
I don't think this can be fixed in the SPIR-V compilation step, it's an API-level problem. Perhaps glslang could add a mode that will effectively append `gl_Position.y = -gl_Position.y` to the end of vertex shaders, but someone pointed out to me that doing so interacts badly with anyone using geometry/tessellation shaders. It's not a magic bullet for everyone and probably isn't worth going into at the glslang level.

---

To go into detail it's not the shader co-ordinate system that's different, it's the NDC space/the NDC to window transform that is different from D3D.

D3D and Vulkan both use an upper-left origin, but while D3D has a left-handed NDC space with y pointing up (on the left), Vulkan uses a right-handed NDC with y pointing down (on the right).

![image](https://cloud.githubusercontent.com/assets/661798/18157314/ebd51b0e-701c-11e6-972b-c9566c6ce3c9.png)
_Diagram from [Matthew Wellings](https://matthewwellings.com/blog/the-new-vulkan-coordinate-system/)_

The difference as I understand it is in the viewport transforms. In [Chapter 23.5](https://www.khronos.org/registry/vulkan/specs/1.0/xhtml/vkspec.html#vertexpostproc-viewport) the viewport transform is straightforward, essentially `pixel.xy = viewport_center.xy + viewport_size.xy * NDC.xy`. That means that because the window co-ordinate system is upper left with y increasing downwards, the NDC also has y pointing downwards to match. D3D [has an implicit flip at this step](https://msdn.microsoft.com/en-us/library/windows/desktop/bb206341(v=vs.85).aspx#vpscale), which means that the window also has y pointing downwards but the NDC has y pointing up.

You can see the effects of this if you look at the [LunarG tri demo](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/demos/tri.c). The code defines vertices for the triangle like so:

```cpp
    const float vb[3][5] = {
        /*      position             texcoord */
        { -1.0f, -1.0f,  0.25f,     0.0f, 0.0f },
        {  1.0f, -1.0f,  0.25f,     1.0f, 0.0f },
        {  0.0f,  1.0f,  1.0f,      0.5f, 1.0f },
    };
```

And the vertex shader is a simple pass-through:

```cpp
layout (location = 0) in vec4 pos;
layout (location = 1) in vec2 attr;
layout (location = 0) out vec2 texcoord;
out gl_PerVertex {
        vec4 gl_Position;
};
void main() {
   texcoord = attr;
   gl_Position = pos;
}
```

And this leads to a triangle pointing down - the 'base' is at `y=-1.0`, the 'point' at `y=1.0`:

![image](https://cloud.githubusercontent.com/assets/661798/18157572/975d883e-701e-11e6-8f5d-610e22492fa8.png)

Hopefully Vulkan's behavior can be fixed to precisely match D3D's with an API extension that makes its way into core. AFAIK this has been how things have worked for a long time and the earliest demos like [the LunarG cube VS](https://github.com/KhronosGroup/Vulkan-LoaderAndValidationLayers/blob/master/demos/cube.vert) have code along the lines of

```cpp
   // GL->VK conventions
   gl_Position.y = -gl_Position.y;
   gl_Position.z = (gl_Position.z + gl_Position.w) / 2.0;
```

at the end of main() to work around it where the projection matrices haven't been changed.

**Comment by Anteru** [comment]
I also think the easiest is to just change the NDC transform to swap y, which is part of the viewport state, instead of forcing it into SPIR-V. This will then match the D3D convention and seems least intrusive to me.

**Comment by xorgy** [comment]
I agree that this doesn't belong in the shader language to begin with, unless I'm missing something.

**Comment by dankbaker** [comment]
This could belong here. SPIRV has a decoration for coordinate system, which I was inquiring if that should be changed for HLSL shaders, if that decoration was what controlled the viewport projection. So that's the first question.  The NDC transformation is probably injected code into the pipline object so this likely is something that is set at pipeline creation time.  It could be argued it's a property of the Fragmentshader, and indeed there appears to be decorations for that.  

However, since it's not allowed on Vulkan, there would have to be a spec change at least to let this happen.

**Comment by johnkslang** [comment]
The SPIR-V decoration **OriginUpperLeft** is for fragment stage only, and I don't think can be used to correct or compensate for an inverted vertex stage.

A quote from the GLSL specification:

> Redeclaring gl_FragCoord with origin_upper_left and/or pixel_center_integer qualifiers only affects
gl_FragCoord.x and gl_FragCoord.y. It has no effect on rasterization, transformation, or any other part
of the OpenGL pipeline or language features.

**Comment by johnkslang** [comment]
I think this is a valid issue, but not against glslang, so am unsure what to do about it.  Maybe it should be logged against Vulkan somehow?

The work around may indeed be to have vertex shaders negate *y*, but should that be in the shader source, rather than in the compiler?

**Comment by jims** [comment]
Doesn't this need to be solved at the API level though? Face culling hardware will need to know about this right? Adding it to spirv will still require applications to invert the triangle winding order compared to what a D3D renderer would use.

**Comment by johnkslang** [comment]
I transferred this to https://github.com/KhronosGroup/Vulkan-Docs/issues/357.
